### PR TITLE
Updated UI, Added Exception Checking, CustomTextWatcher for all constrained input fields

### LINKED ISSUE #189: Update Final UI
- [x] #192
- [x] #193
- [x] #194
- [x] #195
- [x] #205
- [x] #219
- [x] #223
- [x] #224

### LINKED ISSUE #192: Change Font size and Colour

### LINKED ISSUE #217: Change image buttons to text buttons for espresso

### LINKED ISSUE #219: Clean view habit event

### LINKED ISSUE #221: Comments on Project part 3
# Comments and marks on project part 3. Please make issues with respect to comments if needed and link them with this issue.
Do not repeat the same mistakes again, noted by your TA during the meetings. Ask questions as early as possible. Use this issue for asking questions regarding comments etc.

****

### Comment: 
Good - it seems like half requirements are done but the code is a bit hard to follow because of a lack of comments. The are also confusing error messages when attempting to create an account (The app crashes if you don't have a user name but input a password on acocunt creation, and isn't clear you need letters and numbers for a password). The naming conventions are also inconsistent (usually Java is CamalCaseFor class names as opposed_to_using_underscores). 

### Comment:
Good - Some portions of the code are seemingly well documented with clear descriptions and other portions have none at all. 

### Comment:
Good work, made use of using sub tasks for larger stories.

### Comment:
More than one UI Tests fail (e.g. editHabitEvent, deleteHabitEvent, addHabit, removeHabit). Your Unit tests should be more comprehensive in testing model classes. Please remove commented tests or tests that are empty.

### Comment:
In code, divide classe into proper packages. UML classes cant be detached from the rest of the diagram. Layers, MVC, missing. Why can HabitList exist without Habits? USe camel case for class names. Class names cant be verbs, verbs are actions/methods/functions. Wrong identification of aggregations, all classes associated to Habits should be destroyed when a habit destroys. Profile keeps habits right? Then why is is detached. Location missing. Firebase Storage is na inappropriate class name a class name should exhibit its behaviour i.e., what it does. Worng notation for interface implementation used. Instead of udnerscores in class names, use camelCase. A weird unknown circle dot used within UML. UML needs reconstruction.

### Comment:
It seems like all comments were addressed and they were tracked.

### Comment:
Mostly Consistent and clear UI, stroyboard is clear. Need some improvements regarding the intuitiveness of the UI, It is too dense. Reflect the edits done in the app reflect on UI mockups and storyboard. I pointed somethings during the demo, reflect those in the mockups and storyboard as well.

### Comment:
There is some kind of sprint planning. They displays the understanding of scrum. Sprints are not planned by user stories, no priority planning in addressing user stories, No review of sprint shown. user stories are somewhat completed.

### Comment:
Comments and suggestions given on the spot should not be repeated. The organization of presented features should be improved. Prepare well before the demo time, and practice again and again.

### Comment:
You need to improve yourselves in using the tools. You still lack in following some standards.

### Comment:
Somewhat above average relative quality, overall comprehensiveness, creativity, attractiveness, and innovation

****
****

Issues:

- [x] #228
- [x] #229
- [x] #230
- [x] #231
- [x] #232

### LINKED ISSUE #228: Create text watchers for all text input fields with proper input constraints

### LINKED ISSUE #239: Unit Tests
This issue lists the required classes that need to be tested with unit testing.

## Tasks
- [x] #252
- [x] #253
- [x] #254
- [x] #255
- [x] #256
- [x] #257

### LINKED ISSUE #252: Follower

### LINKED ISSUE #253: Following

### LINKED ISSUE #254: Habit

### LINKED ISSUE #255: HabitEvent

### LINKED ISSUE #256: Request

### LINKED ISSUE #257: User

### PR DESCRIPTION
**Pull request by Sohaib404**
changed delete button to text
cleaned up sign up
edit texts now have checks for constrained input
exception checking for bad input to habits and habit events
closes #228 
closes #219 
closes #217 
closes #192 

****

event list item spacing
fix "view habit event"
"remove habit" button changed into red text and made into neutral button
added cancel button on "editHabit"
updated others profile

closes #221 
closes #189 

****
Unit Tests:

closes #252
closes #253
closes #254
closes #255
closes #256
closes #257
closes #239

### CODE REVIEW

**Review by the-rooster — APPROVED** [review]
excellent work dearest sohaib

**Comment by Sohaib404** [comment]
I did it
